### PR TITLE
do not invoke deletion in creation events

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards https://github.com/giantswarm/giantswarm/issues/1526. See also https://github.com/giantswarm/operatorkit/pull/23.

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -15,6 +15,6 @@ test:
 
 deployment:
   master:
-    branch: master
+    branch: last
```
Comment (xh3b4sd) [inline circle.yml]: Has to be fixed before merge.

File: glide.yaml [hunk 2]
```diff
@@ -2,6 +2,7 @@ package: github.com/giantswarm/ingress-operator
 import:
 - package: github.com/giantswarm/microkit
 - package: github.com/giantswarm/operatorkit
+  version: last
```
Comment (xh3b4sd) [inline glide.yaml]: Waiting for approval of the PR.

File: service/operator/operatorkit.go [hunk 3]
```diff
@@ -1,145 +0,0 @@
-package operator
-
-import (
-	microerror "github.com/giantswarm/microkit/error"
-)
-
-// Operator implements the building blocks of any operator business logic being
-// reconciled when observing TPRs. This interface provides a guideline for an
-// easier way to follow the rather complex intentions of operators in general.
-type Operator interface {
```
Comment (xh3b4sd) [inline service/operator/operatorkit.go]: I was debugging and deploying back and forth and wondered why the changes in the vendored dependency are not applied. Well, reason was the operator was still using this code here. I wasted two hours finding this out. Great.

File: service/operator/service.go [hunk 4]
```diff
@@ -164,7 +165,9 @@ func (s *Service) addFunc(obj interface{}) {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
-	err := ProcessCreate(obj, s)
+	s.logger.Log("debug", "executing the operator's addFunc")
```
Comment (xh3b4sd) [inline service/operator/service.go]: I added a lot of logging. Helps getting into the actions.

File: service/operator/service.go [hunk 5]
```diff
@@ -164,7 +165,9 @@ func (s *Service) addFunc(obj interface{}) {
...
+
+	err := operator.ProcessCreate(obj, s)
```
Comment (xh3b4sd) [inline service/operator/service.go]: We finally use the vendored dependency.

File: service/operator/service.go [hunk 6]
```diff
@@ -207,7 +212,13 @@ func (s *Service) GetCurrentState(obj interface{}) (interface{}, error) {
 			if err != nil {
 				return nil, microerror.MaskAny(err)
 			}
+			// Ensure that the map is assignable. This prevents panics down the road
+			// in case the config-map has no data at all.
+			if len(k8sConfigMap.Data) == 0 {
+				k8sConfigMap.Data = map[string]string{}
+			}
```
Comment (xh3b4sd) [inline service/operator/service.go]: We initially set the empty map in `GetCreateState`, but overwrote content sometimes. This is now here to ease the work of making up the create-state. Feels also more right to make up and initialized everything around the current state here in `GetCurrentState`.

File: service/operator/service.go [hunk 7]
```diff
@@ -387,23 +402,25 @@ func (s *Service) GetDeleteState(obj, currentState, desiredState interface{}) (i
 		{
 			newData := map[string]string{}
 			for k, v := range deleteState.ConfigMap.Data {
-				if inConfigMapData(dState.ConfigMapData, k, v) {
+				if !inConfigMapData(dState.ConfigMapData, k, v) {
```
Comment (xh3b4sd) [inline service/operator/service.go]: This changes the way the delete operation is initialized. We do not receive the zero value anymore, but the desired state, which represents the state desired to be deleted. Feels cleaner this way.

Comment (kopiczko) [inline service/operator/service.go]: I think it won't work in flannel-operator. At least it would be not easy to use. Let's say we have etcd entry to be removed. Then delete state should contain all entries but those to be removed, correct? If so then this makes `ProcessDeleteState` unnecessary complicated and error prone.

Comment (kopiczko) [inline service/operator/service.go]: In other places I see `ConfigMap.Data` (dot). Is there a reason why this is inconsistent?

Comment (xh3b4sd) [inline service/operator/service.go]: This would indeed not make sense. In case of the Flannel operator `GetDeleteState` would contain the items to be deleted. `ProcessDeleteState` would then iterate over the items and just remove them. The reason why we have to prepare the whole set here is that we have to update data of an existing resource.

File: service/operator/service.go (hunk 6 above)
Comment (xh3b4sd) [inline service/operator/service.go]: Done.

File: service/operator/service.go (hunk 7 above)
Comment (xh3b4sd) [inline service/operator/service.go]: `foo.ConfigMap.Data` refers to the data of a real config map. `foo.ConfigMapData` refers to config map data that we make up. This depends on the types being used. Check e.g. `state.go`.

Comment (kopiczko) [inline service/operator/service.go]: Ok. What worries me is that we'll have a lot of nuances in state and it may be difficult to track.

Can you describe how this framework can be applied to following code?

```
err := storage.Delete(KeyFromTPO(obj))
if err != nil && !storage.IsNotExist(err) {
    return err
}
```

Comment (xh3b4sd) [inline service/operator/service.go]: The result from `KeyFromTPO(obj)` will be put into some structure. This structure will be returned by `GetDeleteState`. This structure will then be passed to `ProcessDeleteState` and the rest of your code is done in there.
